aws/ec2.go

```diff
@@ -0,0 +1,225 @@
+	}
+
+	checkVal := !checkSourceDestinationValueIsDisable(check)
+	err = ec2Cli.setEC2SourceDestinationCheck(ctx, ec2NetId, checkVal)
+	if err != nil {
+		return fmt.Errorf("error setting src-dst-check for network-interface-id: %s", convertError(err))
+	}
+
+	log.Infof("Successfully set source-destination-check to %t on network-interface-id: %s", checkVal, ec2NetId)
+	return nil
+}
+
+type ec2MetadataClient struct {
+	ec2MetadataSvc *ec2metadata.EC2Metadata
+}
+
+func NewEc2MetadataClient(awsSession *awssession.Session) (*ec2MetadataClient, error) {
+	svc := ec2metadata.New(awsSession)
+	if svc == nil {
+		return nil, fmt.Errorf("error connecting to EC2 Metadata service")
+	}
+
+	return &ec2MetadataClient{
+		ec2MetadataSvc: svc,
+	}, nil
+}
+
+func (c *ec2MetadataClient) getEC2InstanceID(ctx context.Context) (string, error) {
+	idDoc, err := c.ec2MetadataSvc.GetInstanceIdentityDocumentWithContext(ctx)
+	if err != nil {
+		return "", err
+	}
+	log.Debugf("ec2-instance-id: %s", idDoc.InstanceID)
+	return idDoc.InstanceID, nil
+}
+
+func (c *ec2MetadataClient) getEC2Region(ctx context.Context) (string, error) {
+	region, err := c.ec2MetadataSvc.RegionWithContext(ctx)
+	if err != nil {
+		return "", err
+	}
+	log.Debugf("region: %s", region)
+	return region, nil
+}
+
+type ec2Client struct {
+	ec2Svc        *ec2.EC2
+	ec2Region     string
+	ec2InstanceId string
+}
+
+func NewEc2Client(awsSession *awssession.Session, region, instanceId string) (*ec2Client, error) {
+	ec2Svc := ec2.New(awsSession, aws.NewConfig().WithRegion(region))
+	if ec2Svc == nil {
+		return nil, fmt.Errorf("error connecting to EC2 service")
+	}
+
+	return &ec2Client{
+		ec2Svc:        ec2Svc,
+		ec2Region:     region,
+		ec2InstanceId: instanceId,
+	}, nil
+}
+
+func (c *ec2Client) getEC2NetworkInterfaceId(ctx context.Context) (networkInstanceId string, err error) {
+	input := &ec2.DescribeInstancesInput{
+		InstanceIds: []*string{
+			aws.String(c.ec2InstanceId),
+		},
+		Filters: []*ec2.Filter{
+			{
+				Name: aws.String("network-interface.attachment.device-index"),
+				Values: []*string{
+					aws.String("0"), // Only device-id-0
+				},
+			},
+		},
+	}
+
+	i := 0
+retry:
+	out, err := c.ec2Svc.DescribeInstancesWithContext(ctx, input)
+	if err != nil {
+		if retriable(err) && i < defaultRetries {
+			log.WithField("instance-id", c.ec2InstanceId).Debug("retrying getting network-interface-id")
+			i++
+			goto retry
+		}
+		return "", err
+	}
+
+	if len(out.Reservations) <= 0 {
+		return "", fmt.Errorf("no network-interface-id found for EC2 instance %s", c.ec2InstanceId)
+	}
+
+	var interfaceId string
+	for _, instance := range out.Reservations[0].Instances {
+		if len(instance.NetworkInterfaces) == 0 {
+			return "", fmt.Errorf("no network-interface-id found for EC2 instance %s", c.ec2InstanceId)
+		}
+		for _, networkInterface := range instance.NetworkInterfaces {
+			if *(networkInterface.Attachment.DeviceIndex) == 0 {
```

Comment:
Is there a specific reason we want to return just the first network index we find? Would we ever want to return a different one?

Reply:
I thought we passed in the device index to the query, so shouldn't it only ever return us a single network interface with device index 0? Do we need this loop at all?

Reply:
Yeah, it doesn't seem to work. Documentation allows returning using the filter but the response contains all the interfaces. Looping is to make sure` device-index = 0` is modified.